spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalPredictor.scala

```diff
@@ -0,0 +1,154 @@
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{Engine, MklBlas}
+
+
+import scala.reflect.ClassTag
+
+object LocalPredictor {
+
+  def apply[T: ClassTag](model: Module[T], weightsBias: Array[Tensor[T]])
+                        (implicit ev: TensorNumeric[T]): LocalPredictor[T] = {
+    new LocalPredictor[T](model, weightsBias)
+  }
+}
+
+class LocalPredictor[T: ClassTag] private[optim](model: Module[T], weightsBias: Array[Tensor[T]])
+                                                (implicit ev: TensorNumeric[T])
+  extends Serializable {
+
+  val logger = LocalValidator.logger
+  private val coreNumber = Engine.coreNumber()
+
+  private val subModelNumber = Engine.getEngineType match {
+    case MklBlas => coreNumber
+    case _ => throw new IllegalArgumentException
+  }
+
+  private val batchPerCore = 4
+
+  def predictClass(dataSet: Array[Sample[T]]): Array[Int] = {
+    val result = predict(dataSet)
+    result.map(output => {
+      val _output = output.toTensor[T]
+      require(_output.dim() == 1, s"Predictor.predictClass:" +
+        s"Only support one sample has one lable, but got ${_output.dim()} label")
+      ev.toType[Int](_output.max(1)._2.valueAt(1))
+    })
+  }
+
+  def predictClass(dataSet: LocalDataSet[MiniBatch[T]]): Array[Int] = {
+    val result = predict(dataSet)
+    result.map(output => {
+      val _output = output.toTensor[T]
+      require(_output.dim() == 1, s"Predictor.predictClass:" +
+        s"Only support one sample has one lable, but got ${_output.dim()} label")
+      ev.toType[Int](_output.max(1)._2.valueAt(1))
+    })
+  }
+
+  def predict(dataSet: LocalDataSet[MiniBatch[T]]): Array[Activity] = {
+    val dataIter = dataSet.data(train = false)
+
+    val workingModels = (1 to subModelNumber).map(_ => {
+      val submodel = model.cloneModule().evaluate()
+      putWeightBias(weightsBias, submodel)
+      submodel
+    }).toArray
+    dataIter.map(batch => {
+      println("Enter map")
+      val stackSize = batch.size() / subModelNumber
+      val extraSize = batch.size() % subModelNumber
+      val parallelism = if (stackSize == 0) extraSize else subModelNumber
+      val start = System.nanoTime()
+      val result = Engine.default.invokeAndWait(
+        (0 until parallelism).map(b =>
+          () => {
+            val offset = b * stackSize + math.min(b, extraSize) + 1
+            val length = stackSize + (if (b < extraSize) 1 else 0)
+            val currentMiniBatch = batch.slice(offset, length)
+            val input = currentMiniBatch.getInput()
+            val output = workingModels(b).forward(input).toTensor[T]
+            output
+          }
+        )
+      )
+      val batchResult = result.flatMap(_.split(1)).map(_.asInstanceOf[Activity])
+      batchResult
+    }).toArray.flatten
+
+  }
+
+  def predict(dataSet: Array[Sample[T]]): Array[Activity] = {
```

Comment:
cc @jenniew  I guess most of the code could be shared between these two over-loading`predict` methods.

Reply:
and also the `predict-class` methods.